### PR TITLE
golang 1.17 fails to parse IPs with leading zeros

### LINKED ISSUE #100895: [go1.17] Guard against stdlib ParseIP/ParseCIDR changes in API validation
#### What happened:

The stdlib ParseIP/ParseCIDR methods reports no error for some unbounded and questionable inputs

* https://github.com/golang/go/issues/30999
* https://github.com/golang/go/issues/43389

golang is considering tightening the permitted formats (disallowing leading 0s, for example).

While generally positive for new data, this could have the effect of making existing persisted data suddenly be considered invalid. That would cause objects to fail validation on updates that did not modify the existing suddenly-invalid fields, and could prevent objects with that data from being able to be successfully deleted (for example, an object with a finalizer that has to be removed via update prior to deletion, and an immutable field validated with ParseIP/ParseCIDR)

A quick sweep of validation files turned up these uses of these methods:

```
egrep -R '\.(ParseIP|ParseCIDR)\(' pkg staging | grep validation | grep -v test
pkg/apis/core/validation/validation.go:			if ip := net.ParseIP(ns); ip == nil {
pkg/apis/core/validation/validation.go:		if ip := net.ParseIP(hostAlias.IP); ip == nil {
pkg/apis/core/validation/validation.go:	ip := net.ParseIP(ipAddress)
pkg/apis/core/validation/validation.go:			if isIP := (net.ParseIP(ingress.IP) != nil); !isIP {
pkg/apis/core/validation/validation.go:			if isIP := (net.ParseIP(ingress.Hostname) != nil); isIP {
pkg/apis/core/validation/validation.go:	_, net, err := net.ParseCIDR(cidr)
pkg/apis/networking/validation/validation.go:			if isIP := (net.ParseIP(ih.Host) != nil); isIP {
pkg/proxy/apis/config/validation/validation.go:	if net.ParseIP(config.BindAddress) == nil {
pkg/proxy/apis/config/validation/validation.go:			if _, _, err := net.ParseCIDR(config.ClusterCIDR); err != nil {
pkg/proxy/apis/config/validation/validation.go:	if ip := net.ParseIP(hostIP); ip == nil {
pkg/proxy/apis/config/validation/validation.go:		if _, _, err := net.ParseCIDR(nodePortAddresses[i]); err != nil {
pkg/proxy/apis/config/validation/validation.go:		if _, _, err := net.ParseCIDR(excludeCIDRs[i]); err != nil {
staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:	if net.ParseIP(value) == nil {
staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:	ip := net.ParseIP(value)
staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:	ip := net.ParseIP(value)
```

Each of those represent an incompatibility risk if golang tightens data accepted by ParseIP / ParseCIDR

#### What you expected to happen:

Data in existing persisted fields to tolerate no-op updates.

/lifecycle frozen
/priority important-longterm

### PR DESCRIPTION
**Pull request by aojea**
/kind bug
/kind cleanup

#### What this PR does / why we need it:

Since golang 1.17, https://github.com/golang/go/issues/30999,  "In both net.ParseIP and net.ParseCIDR reject leading zeros in the dot-decimal notation of IPv4 addresses."

This can cause that previous valid data becomes invalid, so we should guarantee that this doesn't happen.

In addition, since this change in the golang stdlib as associated a security  CVE-2021-29923, we should check

> While you triage those callsites, it would be good to also check if they are affected by the kind of issues that are motivating the change: if you are validating the inputs with Go and then passing the inputs to the OS or to non-Go applications, the two might disagree on the inputs validity or meaning. This can cause issues, which we would love to hear about at security@golang.org or on the issue, to assess the security risk.

#### Which issue(s) this PR fixes:

Fixes #100895

```release-note
Since golang 1.17 both net.ParseIP and net.ParseCIDR rejects leading zeros in the dot-decimal notation of IPv4 addresses.
Kubernetes will keep allowing leading zeros on IPv4 address to not break the compatibility.
IMPORTANT: Kubernetes interprets leading zeros on IPv4 addresses as decimal, users must not rely on parser alignment to not being impacted by the associated security advisory:
CVE-2021-29923 golang standard library "net" - Improper Input Validation of octal literals in golang 1.16.2 and below standard library "net" results in indeterminate SSRF & RFI vulnerabilities.
Reference: https://nvd.nist.gov/vuln/detail/CVE-2021-29923
```

#### Additional information

I've tried to use the golangci-lint integration, but that depends on the gocritic linter that is the one embedding ruleguard.
This makes this more brittle and more complicated than just installing ruleguard and add our own rules.
It's important to mention that ruleguard rules require to import `github.com/quasilyte/go-ruleguard/dsl` that has a BSD-3 License

### CODE REVIEW

**Comment by aojea** [comment]
ok, this is working now
/assign @liggitt @dims @spiffxp @BenTheElder @thockin

File: hack/ruleguard/rules.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// +build ignore
```
Comment (liggitt) [inline hack/ruleguard/rules.go:1]: would putting this in hack/tools/ruleguard keep the dep out of the root go.mod file?

**Review by liggitt — COMMENTED** [review]
> I've tried to use the golangci-lint integration, but that depends on the gocritic linter that is the one embedding ruleguard.

if the golangci-lint integration didn't work out, can those changes be dropped from the PR to clarify what is actually required?

the find/replace of the function calls looked good, only comments I had were around the dependency location

File: build/tools.go [hunk 2]
```diff
@@ -21,6 +21,7 @@ package tools
 
 import (
 	_ "github.com/onsi/ginkgo/ginkgo"
+	_ "github.com/quasilyte/go-ruleguard/dsl"
```
[the remark this answers is not in the document]
Comment (aojea) [reply build/tools.go:24]: I've tried that and a lot of more different combinations, 
I think that ruleguard requires the project importing this module, to dynamically appy the rules, let me do a gently  ping to the project author @quasilyte to see if he can clarify or offer alternatives ,

File: hack/ruleguard/rules.go (hunk 1 above)
Comment (aojea) [reply hack/ruleguard/rules.go:1]: tried that, didn't work

File: hack/verify-ruleguard.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+# Install golangci-lint
+echo 'installing ruleguard '
+pushd "${KUBE_ROOT}/hack/tools" >/dev/null
+  go install github.com/quasilyte/go-ruleguard/cmd/ruleguard
+popd >/dev/null
+
+cd "${KUBE_ROOT}"
+
+echo 'running ruleguard '
+ruleguard -rules hack/ruleguard/rules.go -c 0 ./...
```
Comment (liggitt) [inline hack/verify-ruleguard.sh:47]: does the `./...` only find uses inside k8s.io/kubernetes, or also the staging repos?

**Comment by aojea** [comment]
> if the golangci-lint integration didn't work out, can those changes be dropped from the PR to clarify what is actually required?

the rules assume "k8s.io/utils/net" imported as "netutils", the golangci-lint change uses importas to enforce that alias, I can do it i a follow up pr, but I think that is useful to keep this consistency

I've tried current import alias tooling, but it is hardcoded to the e2e folder, once you remove that restriction it complain in 100s of files and will make the diff much bigger.

File: hack/verify-ruleguard.sh (hunk 3 above)
Comment (aojea) [reply hack/verify-ruleguard.sh:47]: good catch, I've double checked,  not all staging ... I need to figure it that out

**Comment by liggitt** [comment]
> the rules assume "k8s.io/utils/net" imported as "netutils", the golangci-lint change uses importas to enforce that alias, I can do it i a follow up pr, but I think that is useful to keep this consistency

I'm -1 on forcing a particular import alias... is that required to make ruleguard function properly?

**Comment by aojea** [comment]
> > the rules assume "k8s.io/utils/net" imported as "netutils", the golangci-lint change uses importas to enforce that alias, I can do it i a follow up pr, but I think that is useful to keep this consistency
> 
> I'm -1 on forcing a particular import alias... is that required to make ruleguard function properly?

nope, is required if we want to apply the fix automatically, 
if you run  `ruleguard -rules rules.go -fix` it applies the suggestion, that now is replacing `net.ParseIP()` by `netutils.ParseIPSloppy()`
not a big deal, since you already have to add the imports, let me remove those parts and fix the staging problem

File: cmd/kubeadm/app/componentconfigs/kubeproxy.go [hunk 4]
```diff
@@ -17,15 +17,14 @@ limitations under the License.
 package componentconfigs
 
 import (
-	"net"
-
 	clientset "k8s.io/client-go/kubernetes"
 	kubeproxyconfig "k8s.io/kube-proxy/config/v1alpha1"
```
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/kubeproxy.go:21]: move `netutils "k8s.io/utils/net"` bellow this line.

**Comment by neolit123** [comment]
> ParseIPSloppy

i know that wg-naming is disbanding but can sloppy be considered an offensive term? perhaps a group of people can complain about it in the future?

is the NonStrict suffix a better idea?

/wg naming

File: hack/ruleguard/rules.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+// #100895 since golang 1.17, golang/go#30999, "In both net.ParseIP and net.ParseCIDR
+// reject leading zeros in the dot-decimal notation of IPv4 addresses."
+// The legacy ParseIP and ParseCIDR functions are forked and exported as
+// ParseIPSloppy and ParseCIDRSloppy on kubernetes/utils/net/parse.go
+
+func netParseIP(m dsl.Matcher) {
+	m.Match(`net.ParseIP($ip)`).
+		Report("Use k8s.io/utils/net to parse IP addresses. Kubernetes #100895").
+		Suggest(`netutils.ParseIPSloppy($ip)`)
```
[the remark this answers is not in the document]
Comment (liggitt) [reply hack/ruleguard/rules.go:31]: nevermind, answered below

File: cmd/kubeadm/app/phases/certs/renewal/manager_test.go [hunk 6]
```diff
@@ -32,6 +32,7 @@ import (
 	certtestutil "k8s.io/kubernetes/cmd/kubeadm/app/util/certs"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil"
 	testutil "k8s.io/kubernetes/cmd/kubeadm/test"
+	netutils "k8s.io/utils/net"
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/manager_test.go:35]: move `netutils "k8s.io/utils/net"` bellow `certutil "k8s.io/client-go/util/cert"`

File: cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go [hunk 7]
```diff
@@ -32,6 +32,7 @@ import (
 	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
 	pkiutil "k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil"
 	testutil "k8s.io/kubernetes/cmd/kubeadm/test"
+	netutils "k8s.io/utils/net"
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/readwriter_test.go:35]: move `netutils "k8s.io/utils/net"` bellow `"k8s.io/client-go/util/keyutil"`

File: cmd/kubeadm/app/util/config/initconfiguration.go [hunk 8]
```diff
@@ -42,6 +42,7 @@ import (
 	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/config/strict"
 	kubeadmruntime "k8s.io/kubernetes/cmd/kubeadm/app/util/runtime"
+	netutils "k8s.io/utils/net"
```
Comment (neolit123) [inline cmd/kubeadm/app/util/config/initconfiguration.go:45]: move this bellow `"k8s.io/klog/v2"`

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 9]
```diff
@@ -46,6 +46,7 @@ import (
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 	"k8s.io/kubernetes/cmd/kubeadm/app/features"
 	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	netutils "k8s.io/utils/net"
```
Comment (neolit123) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go:49]: move this bellow `"k8s.io/client-go/util/keyutil"`

**Comment by justaugustus** [comment]
> > ParseIPSloppy
> 
> i know that wg-naming is disbanding but can "sloppy" be considered an offensive term? perhaps a group of people can complain about it in the future?
> 
> is the NonStrict suffix a better idea?
> 
> /wg naming

`Loose` or `Tolerant` are potential options.
Slight preference towards `ParseIPTolerant`, which is similar to usage in other packages we use e.g., https://pkg.go.dev/github.com/blang/semver#ParseTolerant

**Comment by liggitt** [comment]
Thanks for the feedback. IMO, a function name with a vaguely negative connotations is a feature here, since using it should be avoided unless required for backwards-compatibility (we wouldn't want to attract new use). Also, for better or worse, it's in committed revisions of k8s.io/utils which has a strict forward-compatible-only policy.

File: hack/verify-ruleguard.sh (hunk 3 above)
Comment (aojea) [reply hack/verify-ruleguard.sh:47]: this applies to all the linters that use modules, staging folders are omitted

**Comment by aojea** [comment]
is not an option not being able to process staging/*
ok, let me try a different approach, hold on your reviews, I'll ping you again once I got it working

File: hack/ruleguard/rules.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (aojea) [reply hack/ruleguard/rules.go:31]: yes

**Comment by justaugustus** [comment]
> Thanks for the feedback. IMO, a function name with vaguely negative connotations is a feature here, since using it should be avoided unless required for backwards-compatibility (we wouldn't want to attract new use). Also, for better or worse, it's in committed revisions of k8s.io/utils which has a strict forward-compatible-only policy.

Makes sense!
/remove-wg naming

**Comment by aojea** [comment]
all the golang tools seems to have issues related to the modules, the don't iterate over all files :shrug: 
```
time find_files |  xargs gofmt -l -r 'net.ParseCIDR(a) -> netutils.ParseCIDRSloppy(a)' | grep stagi | wc

     12      12    1031

real    0m25,520s
user    0m29,464s
sys     0m0,964s
[aojea@juanan kubernetes]$ 
[aojea@juanan kubernetes]$ 
[aojea@juanan kubernetes]$ time grep -r net.ParseIP\( staging/ | wc
     88     557   16079

real    0m0,078s
user    0m0,037s
sys     0m0,041s
```
I think that we can do it in bash :upside_down_face: , with a simple pattern matching

**Comment by eclipseo** [comment]
Does this PR also make the necessary changes to apimachinery, as mentioned in https://github.com/kubernetes/apimachinery/issues/128, too? I can't find it in Files changed view.

**Comment by aojea** [comment]
> Does this PR also make the necessary changes to apimachinery, as mentioned in [kubernetes/apimachinery#128](https://github.com/kubernetes/apimachinery/issues/128), too? I can't find it in Files changed view.

yeah, that was exactly the problem because I had to hold the PR, some files were missing.

I'm almost done, I will update the PR tomorrow, thanks for following up

**Comment by aojea** [comment]
ok, I've tried all the possibilities to make this as simple as possible, but I couldn't find any tool or tools that helped with the refactoring, and there are a lot of files and imports that has to be changed, so I've created a tool https://github.com/aojea/sloppy-netparser

If tests passes, this should be ready to go

**Comment by liggitt** [comment]
looks like some packages need to have k8s.io/utils/net added as a valid import in import-restrictions files

/skip

**Comment by liggitt** [comment]
lgtm otherwise, thanks for picking this up!

**Comment by aojea** [comment]
```
=== RUN   TestVolumeUnmountAndDetachControllerDisabled
W0820 09:05:54.402158   62913 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
I0820 09:05:54.402758   62913 volume_manager.go:291] "Starting Kubelet Volume Manager"
I0820 09:05:54.402955   62913 desired_state_of_world_populator.go:146] "Desired state populator starts to run"
E0820 09:05:54.405877   62913 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1.CSIDriver: unhandled watch: testing.WatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:"", Verb:"watch", Resource:schema.GroupVersionResource{Group:"storage.k8s.io", Version:"v1", Resource:"csidrivers"}, Subresource:""}, WatchRestrictions:testing.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:""}}
I0820 09:05:54.604805   62913 reconciler.go:244] "operationExecutor.AttachVolume started for volume \"vol1\" (UniqueName: \"fake/fake-device\") pod \"foo\" (UID: \"12345678\") " pod="test/foo"
I0820 09:05:54.604984   62913 reconciler.go:157] "Reconciler: start to sync state"
I0820 09:05:54.605064   62913 operation_generator.go:369] AttachVolume.Attach succeeded for volume "vol1" (UniqueName: "fake/fake-device") from node "127.0.0.1" 
I0820 09:05:54.706173   62913 operation_generator.go:587] "MountVolume.WaitForAttach entering for volume \"vol1\" (UniqueName: \"fake/fake-device\") pod \"foo\" (UID: \"12345678\") DevicePath \"/dev/vdb-test\"" pod="test/foo"
I0820 09:05:54.706461   62913 operation_generator.go:597] "MountVolume.WaitForAttach succeeded for volume \"vol1\" (UniqueName: \"fake/fake-device\") pod \"foo\" (UID: \"12345678\") DevicePath \"/dev/sdb\"" pod="test/foo"
I0820 09:05:54.706715   62913 operation_generator.go:630] "MountVolume.MountDevice succeeded for volume \"vol1\" (UniqueName: \"fake/fake-device\") pod \"foo\" (UID: \"12345678\") device mount path \"\"" pod="test/foo"
==================
WARNING: DATA RACE
Write at 0x00c0004d6a68 by goroutine 20:
  k8s.io/kubernetes/pkg/kubelet.TestVolumeUnmountAndDetachControllerDisabled()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_volumes_test.go:319 +0xc73
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Previous read at 0x00c0004d6a68 by goroutine 104:
  k8s.io/kubernetes/pkg/kubelet.(*fakePodWorkers).ShouldPodRuntimeBeRemoved()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernete
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/104368/pull-kubernetes-unit/1428640022386446336
/test pull-kubernetes-unit

**Comment by aojea** [comment]
> /hold
> [#104368 (comment)](https://github.com/kubernetes/kubernetes/pull/104368#discussion_r691257310)

/hold cancel

**Comment by thockin** [comment]
Thanks @aojea !

FWIW I have ruleguard working EXCEPT across modules.  It can run it from the k/k root and it finds errors in k/k and in vendored modules, but not in staging.  In fact, I have it stored in hack/tools (which is a module) and it works against the root module (presumable because it is a path prefix) but it does not work in staging, still.

This is one of those "go modules is dumb" things but also "ruleguard is dumb".   I filed some issues with ruleguard and golangci-lint, and until those are resolved the only answers I see are:

a) keep using your custom thing
b) symlink EVERYTHING from staging into vendor (which seems to be true today) and just rely on that

**Pull request merged**